packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -895,188 +895,253 @@ class TextPainter {
-    return _caretMetrics.fullHeight;
-  }
+  // Holds the TextPosition the last caret metrics were computed with. When new
+  // values are passed in, we recompute the caret metrics only as necessary.
+  TextPosition? _previousCaretPosition;
 
   // Cached caret metrics. This allows multiple invokes of [getOffsetForCaret] and
   // [getFullHeightForCaret] in a row without performing redundant and expensive
   // get rect calls to the paragraph.
-  late _CaretMetrics _caretMetrics;
+  _CaretMetrics? _cachedCaretMetrics;
 
-  // Holds the TextPosition and caretPrototype the last caret metrics were
-  // computed with. When new values are passed in, we recompute the caret metrics.
-  // only as necessary.
-  TextPosition? _previousCaretPosition;
-  Rect? _previousCaretPrototype;
-
-  // Checks if the [position] and [caretPrototype] have changed from the cached
-  // version and recomputes the metrics required to position the caret.
-  void _computeCaretMetrics(TextPosition position, Rect caretPrototype) {
+  // the metrics required to position the caret.
+  _CaretMetrics? _computeCaretMetrics(TextPosition position) {
     assert(_debugAssertTextLayoutIsValid);
-    if (position == _previousCaretPosition && caretPrototype == _previousCaretPrototype) {
-      return;
+    if (position == _previousCaretPosition) {
+      return _cachedCaretMetrics;
+    }
+    final InlineSpan text = _text!;
+    // Special case: the text is emtpy. ui.Paragraph returns a TextBox located
+    // at the top left corner of the paragraph, regardless of TextAlign or the
+    // writing direction. This doesn't happen when the upstream is a line break
+    // character. Return null in this case without caching.
+    if (text.codeUnitAt(0) == null) {
+      return null;
+    }
+
+    final ui.Paragraph paragraph = _paragraph!;
+
+    // The goal is to search in the direction specified by the affinity, for the
+    // closest grapheme in that direction:
+    //
+    //  * For upstream, the closest grapheme [start, end), where end <= position.
+    //  * For downstream, the closest grapheme [start, end), where start >= position.
+    //
+    // If there is no full grapheme in that direction, try the opposite direction.
+    //
+    // Special case: when the affinity is upstream, and the closest glyph is a
+    // line break, force the search direction to be downstream. This is because
+    // newline characters are reported as zero-width TextBoxes at the **end** of
+    // its line. The behavior of different getBoxesForRange implementations
+    // is not consistent when the input range is [length, length + 1) and the
+    // last character is a line breaker.
+    //
+    // TODO(LongCatIsLooong): introduce ui.Paragraph-level api for this. The
+    // exponential search using getBoxesForRange isn't necessary and does not
+    // work well for certian edge cases (e.g., when the position is inside a
+    // glyph, and the previous character is a newline character).
```

Comment:
Nit: Consider linking/creating a GitHub issue here.

Reply:
This will be taken care of in the new paragraph api. But I don't think there's a github issue.